File: cli/sync.ts
import { CONFIG_DIR, DB_PATH, META_PATH, ensureConfigDir, fileExists, readJson, writeJson } from "../src/utils/fs.ts"
import { rename, unlink } from "node:fs/promises"
import { Database } from "bun:sqlite"

const REPO = "jongjinchoi/ossdive"
const TAG  = "db-latest"
const TTL  = 60 * 60 * 1000  // 1h in ms

interface SyncMeta {
  assetId:   number
  updatedAt: string
  syncedAt:  number
}

interface GHAsset {
  id:                  number
  name:                string
  updated_at:          string
  browser_download_url: string
}

interface GHRelease {
  assets: GHAsset[]
}

export type SyncStatus = "fresh" | "cached" | "updated" | "offline" | "missing"

export function selectDbAsset(assets: GHAsset[]): GHAsset | undefined {
  const versioned = assets
    .filter(a => /^ossdive-[A-Za-z0-9_.-]+\.db$/.test(a.name))
    .sort((a, b) => a.updated_at.localeCompare(b.updated_at))

  return versioned.at(-1) ?? assets.find(a => a.name === "ossdive.db")
}

function apiHeaders(): Record<string, string> {
  const headers: Record<string, string> = { "Accept": "application/vnd.github+json" }
  if (process.env["GITHUB_TOKEN"]) {
    headers["Authorization"] = `Bearer ${process.env["GITHUB_TOKEN"]}`
  }
  return headers
}

async function fetchRelease(): Promise<GHRelease> {
  const url = `https://api.github.com/repos/${REPO}/releases/tags/${TAG}`
  const res  = await fetch(url, { headers: apiHeaders(), signal: AbortSignal.timeout(8000) })
  if (!res.ok) throw new Error(`GitHub API ${res.status}: ${res.statusText}`)
  return res.json() as Promise<GHRelease>
}

async function downloadAsset(assetId: number, dest: string): Promise<void> {
  const url = `https://api.github.com/repos/${REPO}/releases/assets/${assetId}`
  const res  = await fetch(url, {
    headers: { ...apiHeaders(), "Accept": "application/octet-stream" },
    signal:  AbortSignal.timeout(30_000),
  })
  if (!res.ok) throw new Error(`Download failed ${res.status}: ${res.statusText}`)
  const buf = await res.arrayBuffer()
  const tmp = `${dest}.tmp-${process.pid}-${Date.now()}`
  try {
    await Bun.write(tmp, buf)
    assertValidSQLite(tmp)
    await rename(tmp, dest)
  } catch (err) {
    await unlink(tmp).catch(() => {})
    throw err
  }
}

function assertValidSQLite(path: string): void {
  const db = new Database(path, { readonly: true, strict: true })
  try {
    const row = db.prepare("PRAGMA integrity_check").get() as { integrity_check: string }
    if (row.integrity_check !== "ok") {
      throw new Error(`Downloaded DB failed integrity_check: ${row.integrity_check}`)
    }
  } finally {
    db.close()
  }
}

export async function syncDb({ force = false }: { force?: boolean } = {}): Promise<{
  status: SyncStatus
  path:   string
}> {
  // Dev override: OSSDIVE_DB env var bypasses sync entirely (same as MCP server pattern)
  if (process.env["OSSDIVE_DB"]) {
    return { status: "cached", path: process.env["OSSDIVE_DB"] }
  }

  await ensureConfigDir()
  const meta = await readJson<SyncMeta>(META_PATH)

  // TTL cache hit: skip network entirely
  if (!force && meta && Date.now() - meta.syncedAt < TTL && (await fileExists(DB_PATH))) {
    return { status: "cached", path: DB_PATH }
  }

  try {
    const release = await fetchRelease()
    const asset   = selectDbAsset(release.assets)
    if (!asset) throw new Error("ossdive.db asset not found in db-latest release")

    // Same version as cached: update syncedAt timestamp, skip download
    if (!force && meta?.updatedAt === asset.updated_at && (await fileExists(DB_PATH))) {
      await writeJson(META_PATH, { ...meta, syncedAt: Date.now() })
      return { status: "fresh", path: DB_PATH }
    }

    await downloadAsset(asset.id, DB_PATH)
    await writeJson(META_PATH, { assetId: asset.id, updatedAt: asset.updated_at, syncedAt: Date.now() })
    return { status: "updated", path: DB_PATH }
  } catch {
    // Network failure: fall back to local DB if it exists
    if (await fileExists(DB_PATH)) return { status: "offline", path: DB_PATH }
    return { status: "missing", path: DB_PATH }
  }
}

export { CONFIG_DIR, DB_PATH }

File: app/src-tauri/src/sync.rs
use serde::{Deserialize, Serialize};
use std::sync::OnceLock;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::db::{config_dir, db_path, meta_path};

const REPO: &str = "jongjinchoi/ossdive";
const TAG: &str = "db-latest";
const TTL_MS: u64 = 3600 * 1000; // 1h in milliseconds (matches CLI Date.now())

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum SyncStatus {
    Fresh,
    Cached,
    Updated,
    Offline,
    Missing,
}

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
struct SyncMeta {
    asset_id:   u64,
    updated_at: String,
    synced_at:  u64,
}

#[derive(Deserialize)]
struct GhAsset {
    id:         u64,
    name:       String,
    updated_at: String,
}

#[derive(Deserialize)]
struct GhRelease {
    assets: Vec<GhAsset>,
}

fn now_millis() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn read_meta() -> Option<SyncMeta> {
    let bytes = std::fs::read(meta_path()).ok()?;
    serde_json::from_slice(&bytes).ok()
}

fn write_meta(meta: &SyncMeta) {
    if let Ok(json) = serde_json::to_string_pretty(meta) {
        let _ = std::fs::write(meta_path(), json);
    }
}

fn http_client() -> &'static reqwest::Client {
    static CLIENT: OnceLock<reqwest::Client> = OnceLock::new();
    CLIENT.get_or_init(|| {
        reqwest::Client::builder()
            .user_agent("ossdive")
            .build()
            .expect("failed to build HTTP client")
    })
}

macro_rules! or_fallback {
    ($expr:expr, $db:expr) => {
        match $expr {
            Ok(v) => v,
            Err(_) => return if $db.exists() { SyncStatus::Offline } else { SyncStatus::Missing },
        }
    };
}

pub async fn sync_db() -> SyncStatus {
    // OSSDIVE_DB env bypasses sync (dev override, same as CLI)
    if std::env::var("OSSDIVE_DB").is_ok() {
        return SyncStatus::Cached;
    }

    let db   = db_path();
    let meta = tauri::async_runtime::spawn_blocking(read_meta)
        .await
        .unwrap_or(None);

    // TTL cache hit
    if let Some(ref m) = meta {
        if now_millis().saturating_sub(m.synced_at) < TTL_MS && db.exists() {
            return SyncStatus::Cached;
        }
    }

    let client = http_client();

    let url = format!(
        "https://api.github.com/repos/{}/releases/tags/{}",
        REPO, TAG
    );

    let resp = or_fallback!(
        client
            .get(&url)
            .header("Accept", "application/vnd.github+json")
            .send()
            .await,
        db
    );
    let resp = or_fallback!(resp.error_for_status(), db);
    let release: GhRelease = or_fallback!(resp.json().await, db);

    let asset = match release.assets.into_iter().find(|a| a.name == "ossdive.db") {
        Some(a) => a,
        None => return if db.exists() { SyncStatus::Offline } else { SyncStatus::Missing },
    };

    // Same version as cached — update synced_at only
    if let Some(ref m) = meta {
        if m.updated_at == asset.updated_at && db.exists() {
            let new_meta = SyncMeta {
                asset_id:   m.asset_id,
                updated_at: m.updated_at.clone(),
                synced_at:  now_millis(),
            };
            let _ = tauri::async_runtime::spawn_blocking(move || write_meta(&new_meta)).await;
            return SyncStatus::Fresh;
        }
    }

    // Download to .tmp, then atomic rename
    let dl_url = format!(
        "https://api.github.com/repos/{}/releases/assets/{}",
        REPO, asset.id
    );

    let resp = or_fallback!(
        client
            .get(&dl_url)
            .header("Accept", "application/octet-stream")
            .send()
            .await,
        db
    );
    let resp  = or_fallback!(resp.error_for_status(), db);
    let bytes = or_fallback!(resp.bytes().await, db);

    let tmp      = config_dir().join("ossdive.db.tmp");
    let db_clone = db.clone();
    let new_meta = SyncMeta {
        asset_id:   asset.id,
        updated_at: asset.updated_at,
        synced_at:  now_millis(),
    };

    let result = tauri::async_runtime::spawn_blocking(move || -> std::io::Result<()> {
        std::fs::create_dir_all(config_dir())?;
        std::fs::write(&tmp, &bytes)?;
        std::fs::rename(&tmp, &db_clone)?;
        write_meta(&new_meta);
        Ok(())
    })
    .await;

    match result {
        Ok(Ok(())) => SyncStatus::Updated,
        _ => if db.exists() { SyncStatus::Offline } else { SyncStatus::Missing },
    }
}
